cloud/services/virtualmachines/virtualmachines.go

```diff
@@ -212,6 +224,63 @@ func (s *Service) Delete(ctx context.Context, spec interface{}) error {
 	return err
 }
 
+func (s *Service) getAddresses(ctx context.Context, vm compute.VirtualMachine) ([]corev1.NodeAddress, error) {
+
+	addresses := []corev1.NodeAddress{}
+
+	for _, nicRef := range *vm.NetworkProfile.NetworkInterfaces {
+
+		// The full ID includes the name at the very end. Split the string and pull the last element
+		// Ex: /subscriptions/$SUB/resourceGroups/$RG/providers/Microsoft.Network/networkInterfaces/$NICNAME
+		explodedNicID := strings.Split(to.String(nicRef.ID), "/")
+		nicName := explodedNicID[len(explodedNicID)-1]
+
+		// Fetch nic and append its addresses
+		nicInterface, err := networkinterfaces.NewService(s.Scope).Get(ctx, &networkinterfaces.Spec{Name: nicName})
+		if err != nil {
+			return addresses, err
+		}
+		nic, ok := nicInterface.(network.Interface)
+		if !ok {
+			return addresses, errors.New("unexpected type for network interface")
+		}
+
+		for _, ipConfig := range *nic.IPConfigurations {
+			if ipConfig.PrivateIPAddress != nil {
+				addresses = append(addresses,
+					corev1.NodeAddress{
+						Type:    corev1.NodeInternalIP,
+						Address: to.String(ipConfig.PrivateIPAddress),
+					},
+				)
+			}
+
+			if ipConfig.PublicIPAddress != nil {
```

Comment:
You can directly use

ipConfig.PublicIPAddress.IPAddress

Reply:
I thought so too and actually implemented it that way initially. However, I found that for some reason, the only populated field in `PublicIPAddress` was the ID. All other fields were empty. This led me down this path of fetching that PublicIPAddress with the ID as it's implemented now.

Reply:
Interesting. Thank you for the explanation, will try and followup with internal teams

Reply:
I actually ran into something similar just yesterday (I think) while creating a Talos cluster manually in Azure. May be another data point to mention.

I think there's a difference in the behavior between public IP addresses that are a basic sku type vs the standard sku type. In the case yesterday, there were no values for public IP that were returned when querying the metadata service. But switching back to the basic sku type, they metadata had all the necessary info.

Reply:
@rsmitty long time no see :)

@awesomenix as of 8/29/2019, SLB team confirmed Standard SKU public IP per instance is incompatible with VMSS and SLB. I’m surprised Spencer didn’t see provisioning errors, as this should trigger them. 

Of course, it’s possible they’ve changed that in the last couple months.